packages/xen-api/src/index.js

```diff
@@ -1004,6 +1008,53 @@ export class Xapi extends EventEmitter {
       delete taskWatchers[ref]
     }
   }
+  async _findCurrentMaster() {
+    debug('%s: will try to find a new master', this._humanId)
+    const hosts = await this.getAllRecords('host')
+    const hostAddresses = this.fallBackAddresses.concat(hosts.map(({ address }) => address))
+
+    // there can be a master only if there is more than 1 host
+    if (hostAddresses.length < 2) {
+      return Promise.resolve()
+    }
+    // if we were already connected
+    for (const hostAdress of hostAddresses) {
+      // not current master send him a simple query and
+      if (hostAdress !== this.url.hostname) {
+        continue
+      }
+      // @todo handle _reverseHostIpAddresses
+      const url = { ...this.url, hostname: hostAdress }
+      // instantiate a transport to this url
+      const transport = autoTransport({
+        secureOptions: {
+          minVersion: 'TLSv1',
+          rejectUnauthorized: !this._allowUnauthorized,
+        },
+        url,
+        httpProxy: this._httpProxy,
+      })
+
+      try {
+        // will not succeed without sessionId, but I only want to know if this host is entitled to answer (he's the master)
+        await transport('session.get_auth_user_name', [])
+      } catch (error) {
+        if (error.code === 'SESSION_INVALID') {
+          debug('%s: found new master %s master', this._humanId, hostAdress)
+          // lucky : we found the master with the first host answering
+          this._setUrl({ ...this._url, hostname: hostAdress })
+          break // we found the master no need to check for the other slave
+        }
+        if (error.code === 'HOST_IS_SLAVE') {
+          debug('%s: slave %s told us the master is %s', this._humanId, hostAdress, error.params[0])
+          this._setUrl({ ...this._url, hostname: error.params[0] })
+          break // we found the master no need to check for the other slave
+        }
+        // any other error should be reported
+        throw error
+      }
```

Comment:
Finding the `master` is not necessary, we are only trying to find a reachable host and if not a `slave`, it will redirect us.

Reply:
that's my understanding, too
If the first fallback is the new master, I will use this URL, if not, I will use the url returned

the only case where I'm making multiple queries is when the fallback host are down